core/src/server/rpc_module/types.rs

```diff
@@ -0,0 +1,389 @@
+		}
+	}
+}
+
+impl<'a> std::fmt::Debug for ConnState<'a> {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		f.debug_struct("ConnState").field("conn_id", &self.conn_id).finish()
+	}
+}
+
+pub(crate) type Subscribers = Arc<Mutex<FxHashMap<SubscriptionKey, (MethodSink, mpsc::Receiver<()>)>>>;
+
+/// This represent a response to a RPC call
+/// and `Subscribe` calls are handled differently
+/// because we want to prevent subscriptions to start
+/// before the actual subscription call has been answered.
+#[derive(Debug, Clone)]
+pub enum CallOrSubscription {
+	/// The subscription callback itself sends back the result
+	/// so it must not be sent back again.
+	Subscription(SubscriptionAnswered),
+
+	/// Treat it as ordinary call.
+	Call(MethodResponse),
+}
+
+impl CallOrSubscription {
+	/// Extract the JSON-RPC response.
+	pub fn as_response(&self) -> &MethodResponse {
+		match &self {
+			Self::Subscription(r) => match r {
+				SubscriptionAnswered::Yes(r) => r,
+				SubscriptionAnswered::No(r) => r,
+			},
+			Self::Call(r) => r,
+		}
+	}
+
+	/// Convert the `CallOrSubscription` to JSON-RPC response.
+	pub fn into_response(self) -> MethodResponse {
+		match self {
+			Self::Subscription(r) => match r {
+				SubscriptionAnswered::Yes(r) => r,
+				SubscriptionAnswered::No(r) => r,
+			},
+			Self::Call(r) => r,
+		}
+	}
+}
+
+/// A complete subscription message or partial subscription message.
+#[derive(Debug, Clone)]
+pub enum SubscriptionMessageInner {
+	/// Complete JSON message.
+	Complete(String),
+	/// Need subscription ID and method name.
+	NeedsData(String),
+}
+
+/// Subscription message.
+#[derive(Debug, Clone)]
+pub struct SubscriptionMessage(pub(crate) SubscriptionMessageInner);
+
+impl SubscriptionMessage {
+	/// Create a new subscription message from JSON.
+	pub fn from_json(t: &impl Serialize) -> Result<Self, serde_json::Error> {
+		serde_json::to_string(t).map(|json| SubscriptionMessage(SubscriptionMessageInner::NeedsData(json)))
+	}
+
+	pub(crate) fn from_complete_message(msg: String) -> Self {
+		SubscriptionMessage(SubscriptionMessageInner::Complete(msg))
+	}
+
+	pub(crate) fn empty() -> Self {
+		Self::from_complete_message(String::new())
+	}
+}
+
+/// Represent a unique subscription entry based on [`RpcSubscriptionId`] and [`ConnectionId`].
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) struct SubscriptionKey {
+	pub(crate) conn_id: ConnectionId,
+	pub(crate) sub_id: RpcSubscriptionId<'static>,
+}
+
+/// Callback wrapper that can be either sync or async.
+#[derive(Clone)]
+pub enum MethodKind {
+	/// Synchronous method handler.
+	Sync(SyncMethod),
+	/// Asynchronous method handler.
+	Async(AsyncMethod<'static>),
+	/// Subscription method handler.
+	Subscription(SubscriptionMethod<'static>),
+	/// Unsubscription method handler.
+	Unsubscription(UnsubscriptionMethod),
+}
+
+/// Method callback wrapper that contains a sync or async closure,
+#[derive(Clone, Debug)]
+pub struct MethodCallback {
```

Comment:
What's the purpose of this? It adds no functionality over `MethodKind`

Reply:
it's simplify a convenience wrapper of methods calls, it could be removed if it helps but it's quite readable is it I reckon.

Reply:
I consider it is unnecessary abstraction and only makes the code harder to reason. The less code, the easier to understand.